### PR TITLE
feat: add persistence migration utility for legacy V1 SDK compatibility

### PR DESCRIPTION
**Pull request by SKannaniOS**
## Description
This PR introduces a comprehensive persistence migration system that enables seamless transition from the legacy RudderStack iOS SDK (V1) to the new Swift SDK. The migration utility reads legacy persistence data from both plist files and UserDefaults, then stores it in the new Swift SDK's scoped UserDefaults format.

The implementation provides backwards compatibility for existing users upgrading their SDK version, ensuring that crucial user data like anonymous IDs, user IDs, traits, and session information are preserved during the migration process.

## Type of change
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Code refactor/optimization

## Implementation Details
- **Added `MigrationUtils.swift`**: Utility class containing helper methods for:
  - JSON encoding/decoding operations
  - File system operations for reading legacy plist files
  - UserDefaults operations for both legacy and new SDK formats
  - Timestamp conversion from wall-clock time to system uptime
  - Cleanup operations for legacy data removal
  
- **Added `PersistenceMigrator.swift`**: Main migration orchestrator that:
  - Reads legacy data from both plist files (`rsDefaultsPersistence.plist`) and `UserDefaults.standard`
  - Transforms legacy data format to new SDK format
  - Migrates user identity data (anonymous ID, user ID, traits)
  - Migrates session management data (session ID, session tracking state, activity timestamps)
  - Provides comprehensive logging for migration process
  - Cleans up legacy data after successful migration

- **Updated SwiftUIExample Xcode project**: Added the new migration files to the example project for demonstration and testing purposes

### Key Migration Features:
- **Legacy Data Sources**: Supports reading from both plist files and UserDefaults.standard
- **Identity Migration**: Preserves anonymous ID, user ID, and user traits
- **Session Continuity**: Converts legacy session data including timestamp format conversion
- **Data Cleanup**: Removes legacy data after successful migration to prevent conflicts
- **Error Handling**: Graceful fallbacks and detailed logging for troubleshooting
- **Platform Support**: Compatible with iOS, tvOS, and other Apple platforms

## Checklist
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added the necessary documentation (if appropriate).
- [x] I have ensured that my code follows the project's code style.
- [x] I have checked for potential performance impacts and optimized if necessary.
- [x] I have checked the code for security issues.
- [ ] I have updated the changelog (if required).

### Testing Migration Functionality:

1. **Simulate Legacy Data**: 
   - Create legacy persistence data in UserDefaults.standard or rsDefaultsPersistence.plist
   - Use legacy keys: `rl_anonymous_id`, `rl_traits`, `rl_session_id`, etc.

2. **Initialize Migration**:
   ```swift
   let migrator = PersistenceMigrator(writeKey: "your-write-key")
   migrator.restorePersistence()
   ```

3. **Verify Migration**:
   - Check that data appears in the new SDK's scoped UserDefaults
   - Verify legacy data is cleaned up after migration
   - Confirm timestamp conversion for session data works correctly

4. **Test Edge Cases**:
   - No legacy data present (should handle gracefully)
   - Invalid JSON in traits field
   - Timestamp conversion when device was rebooted
   - Missing required fields (e.g., session ID)

### Test Environment:
- iOS Simulator with SwiftUIExample project
- Multiple write keys to test scoped UserDefaults isolation
- Various legacy data scenarios (complete, partial, malformed)

## Breaking Changes
None. This is a new feature that provides backwards compatibility and doesn't affect existing functionality.

## Maintainers Checklist
- [ ] The code has been reviewed.
- [ ] CI tests have passed.
- [ ] All necessary documentation has been updated.

## Screenshots (if applicable)
Not applicable - this is a backend utility feature with no UI components.

## Additional Context
This migration system addresses the need for seamless SDK upgrades without losing user data. The implementation follows these design principles:

- **Fail-Safe**: Migration failures don't crash the app or corrupt existing data
- **One-Time Operation**: Legacy data is cleaned up after successful migration
- **Comprehensive Logging**: Detailed logs help with troubleshooting migration issues
- **Platform Agnostic**: Works across all supported Apple platforms
- **Scoped Storage**: Respects the new SDK's write-key-based data isolation

### Related Context:
- Supports migration from RudderStack iOS SDK V1 to Swift SDK
- Handles the architectural change from global UserDefaults to scoped UserDefaults
- Addresses timestamp format changes between SDK versions
- Provides foundation for future SDK compatibility layers

### Technical Notes:
- Uses Bundle identifier + writeKey for scoped UserDefaults suite names
- Converts legacy wall-clock timestamps to system uptime format
- Handles JSON serialization/deserialization for complex data structures
- Implements defensive programming practices for data integrity

## Summary by CodeRabbit

* **New Features**
  * Added data migration capability to transfer persisted values from legacy iOS SDK V1 to the Swift SDK, enabling seamless data preservation during SDK upgrades.
  * Included example code demonstrating how to enable the migration process.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: Examples/SwiftUIExample/SwiftUIExample/PersistenceMigrator/PersistenceMigrator.swift [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+ let migrator = PersistenceMigrator(writeKey: "writeKey")
+ migrator.restorePersistence()
+ ```
+ */
+final class PersistenceMigrator {
+    
+    // MARK: - Properties
+    
+    /** The write key used to identify the analytics instance */
+    let writeKey: String
```
Comment (1abhishekpandey) [inline Examples/SwiftUIExample/SwiftUIExample/PersistenceMigrator/PersistenceMigrator.swift]: Can we highlight which writeKey we are talking about? Is it the new or old?

File: Examples/SwiftUIExample/SwiftUIExample/PersistenceMigrator/PersistenceMigrator.swift [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+ This method is not thread-safe for concurrent calls. Ensure it is called from a single location (e.g., in
+ `application(_:didFinishLaunchingWithOptions:)` or app initialization).
+ 
+ Usage:
+ ```swift
+ let migrator = PersistenceMigrator(writeKey: "writeKey")
+ migrator.restorePersistence()
+ ```
+ */
+final class PersistenceMigrator {
```
Comment (1abhishekpandey) [inline Examples/SwiftUIExample/SwiftUIExample/PersistenceMigrator/PersistenceMigrator.swift]: Can you update the access specifier of all the methods that are not meant to be used by the user, i.e., add `private` explicitly to those methods?

File: Examples/SwiftUIExample/SwiftUIExample/PersistenceMigrator/FromV1/PersistentMigratorFromV1.swift
Comment (1abhishekpandey) [inline Examples/SwiftUIExample/SwiftUIExample/PersistenceMigrator/FromV1/PersistentMigratorFromV1.swift:1]: We also need to handle `appVersion` and `appBuild`.

Let's write an explicit safeguard mechanism to detect a few edge cases, e.g., if the SDK location already exists, then do nothing! It is important to ensure we don't overwrite the value in the new SDK. 

Look for other similar edge cases.

File: Examples/SwiftUIExample/SwiftUIExample/PersistenceMigrator/PersistenceMigrator.swift [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+    
+    // MARK: - Public API
+    
+    /**
+     Reads legacy persistence data and migrates it to Swift SDK UserDefaults.
+     
+     Checks plist file first, then UserDefaults fallback. Logs migration progress. Clears legacy values after successful migration.
+     Returns without performing migration, if no legacy data found.
+     */
+    func restorePersistence() {
```
Comment (1abhishekpandey) [inline Examples/SwiftUIExample/SwiftUIExample/PersistenceMigrator/PersistenceMigrator.swift]: NIT: We could refactor this method to make it more readable. Divided the work into different phases (i.e., different methods).

File: Examples/SwiftUIExample/SwiftUIExample/PersistenceMigrator/PersistenceMigrator.swift (hunk 2 above)
Comment (SKannaniOS) [reply Examples/SwiftUIExample/SwiftUIExample/PersistenceMigrator/PersistenceMigrator.swift]: Any methods that aren’t meant to be used by users are already kept in a private extension of this class.

Comment (1abhishekpandey) [reply Examples/SwiftUIExample/SwiftUIExample/PersistenceMigrator/PersistenceMigrator.swift]: This [readPersistence()](https://github.com/rudderlabs/rudder-sdk-swift/blob/6836b4aab77c6f4b5fae6455723b05938e47b182/Examples/SwiftUIExample/SwiftUIExample/PersistenceMigrator/PersistenceMigrator.swift#L109C10-L109C27) is public but not properly documented. Can we please update its access specifier to private (as I can see it's already used in the [restorePersistence()](https://github.com/rudderlabs/rudder-sdk-swift/blob/6836b4aab77c6f4b5fae6455723b05938e47b182/Examples/SwiftUIExample/SwiftUIExample/PersistenceMigrator/PersistenceMigrator.swift#L53)) or document how to use this method at the [top](https://github.com/rudderlabs/rudder-sdk-swift/blob/6836b4aab77c6f4b5fae6455723b05938e47b182/Examples/SwiftUIExample/SwiftUIExample/PersistenceMigrator/PersistenceMigrator.swift#L20-L25)?

As of now, there are two public methods in this class and it's not clear which one should be used and in what order.

Comment (SKannaniOS) [reply Examples/SwiftUIExample/SwiftUIExample/PersistenceMigrator/PersistenceMigrator.swift]: Both methods are independent to each other. I added documentations for that.

File: Examples/SwiftUIExample/SwiftUIExample/SwiftUIExampleApp.swift [hunk 4]
```diff
@@ -33,6 +33,9 @@ class AppDelegate: NSObject, UIApplicationDelegate {
     
     func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey : Any]? = nil) -> Bool {
         
+        // Restore persistence from legacy storage if exists.
+        PersistenceMigrator(writeKey: "swift-sdk-write-key").restorePersistence()
```
Comment (1abhishekpandey) [inline Examples/SwiftUIExample/SwiftUIExample/SwiftUIExampleApp.swift]: Let's comment this code.


```suggestion
//        PersistenceMigrator(writeKey: "swift-sdk-write-key").restorePersistence()
```

File: Examples/SwiftUIExample/SwiftUIExample/PersistenceMigrator/FromV1/PersistentMigratorFromV1.swift
Comment (1abhishekpandey) [inline Examples/SwiftUIExample/SwiftUIExample/PersistenceMigrator/FromV1/PersistentMigratorFromV1.swift:1]: Let's have the directory and file names like this:
- SwiftUiExample
    - PersistentMigrator
        - FromV1
            - PersistenceMigratorFromV1.swift
            - LegacyDataModelV1.swift
            - MigrationUtilsV1.swift
        - FromV2
            - PersistenceMigratorFromV2.swift
            - LegacyDataModelV2.swift
            - MigrationUtilsV2.swift
    
This is to ensure we provide full visibility on the migration code at a single glance.

**Review by 1abhishekpandey — APPROVED** [review]
